New Post: Debezium Server streaming from MySQL performance (#1213)
debezium/debezium.github.io: _posts/2026-01-08-measuring-debezium-server-performance-mysql-streaming.adoc

More complex architectures should be adopted only when clear scalability, availability or integration requirements justify them. 

In summary, Debezium Server emerges as a simple, efficient and high-performance CDC solution that challenges the assumption that Kafka-centric or Kubernetes-based architectures are always required.
For lightweight projects and streamlined deployments, it offers a compelling balance of maturity, simplicity and performance, delivering real-time change data capture while keeping infrastructure complexity to a minimum. 

[id=follow-up]
== What’s next?

While the results presented above provide a clear view of Debezium Server behavior under default configurations, they also raise additional questions worth exploring in follow-up experiments:

- Discover Debezium Server limits: improve this experiments architecture and configurations to discover what is the point where Debezium Server cannot recover from it.
- Deploy Debezium with Kafka Connect: run this experiment with a distribuited and more complex architecture, using Kafka Connect to deploy Debezium.

> - Get involved with Snapshotting: we have miss one important Debezium feature, and Snapshotting deserves performance investigation.

Reviewer: ```suggestion
- Get involved with snapshotting: we have skiped over one important Debezium feature, and snapshotting deserves performance investigation.
```

- Run different sets of workloads and databases: we have run this experiment against MySQL and used a uniform workload. 

> Include different database, connectors and workload will give us the complete vision of Debezium capabilities. 

Reviewer: ```suggestion

Including of different databases, connectors and workload will give us the complete vision of Debezium capabilities. 
```

- Identify bottlenecks and limiting factors
- Help ensure predictable scalability
- Reduce operational and infrastructure costs
- Deliver improvements in reliability and resilience
- Build trust with users and stakeholders

[id=debezium-server]
== Debezium Server

Apart from performance tests, another great unknown in the CDC community might be *Debezium Server*. If you are reading this, you probably know something about Debezium and CDC.
But let me introduce you to Debezium Server. 
Debezium Server is a standalone runtime for Change Data Capture that captures database changes using Debezium and delivers them directly to configurable sinks without requiring Kafka or Kafka Connect.
It addresses a common gap in CDC architectures: situations where Kafka Connect, Kubernetes, or large distributed platforms introduce unnecessary operational complexity. 
It enables teams to adopt CDC with minimal infrastructure while still benefiting from Debezium’s mature and proven connectors.

++++
<div class="imageblock centered-image">
    <img src="/assets/images/debezium-server-architecture.png" class="responsive-image" alt="Debezium Server Architecture">
</div>
++++

> Debezium Server is built on top of the Debezium Engine, packaging it into a production-ready application that can run as a single JVM process.

Reviewer 2: ```suggestion
Debezium Server is built on top of https://quarkus.io[Quarkus] and the Debezium Engine.
This combination provides a production-ready packaged runtime that can run as a single JVM process.
```

It runs as a single process and although Kafka is not required, it preserves the core CDC guarantees: 

- At-least-once offset management (per connector)
- Ordered event delivery per table (depends on the connector)
- Snapshotting and streaming of the changes in the database
- Rich metadata about transactions and schemas

[id=benchmarking]
== Benchmarking

Now that we are familiar with all the components, let us dive into the technical details. 

Starting with the hardware, we have deployed two EC2 machines in AWS of type "c5a.xlarge". Each instance provides 4 vCPUs and 8 GiB of memory, ensuring sufficient resources for both workload generation and CDC processing.
We have called these two instances _Observed_ and _Observer_ (in order to separate the monitoring load from the machine where Debezium and MySQL are running). In the Observer instance, we have deployed Prometheus, Grafana and YCSB

> (the software piece generating the load). In the Observed instance, we have a single Kafka container (although it is not really necessary and another software can be used),

Reviewer 2: ```suggestion
(the software piece generating the load). In the Observed instance, we have a single Kafka container (although it is not really necessary and could be substituted with any other target, such as Pulsar, Kinesis, PubSub, or even a custom sink solution),
```

In the lower panel, you can see that Debezium's internal queue maintains a high and stable capacity throughout the execution, with no indication of saturation or backlog accumulation.
Overall, the chart confirms that Debezium Server processes MySQL changes efficiently and in near real time, and that the throughput behavior is driven by the database write pattern rather than by downstream processing constraints. 

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/mysql_vs_debezium_server_throughput.png" class="responsive-image" alt="MySQL vs Debezium Server throughput comparison">
</div>
++++

[id=conclusion-recommendations]
== Conclusion and recommendations

This study demonstrates that Debezium Server is a mature, production-ready CDC solution and a credible alternative to traditional CDC architectures based on Kafka Connect, Kubernetes or complex distributed setups.
Throughout all tested scenarios, Debezium Server showed stable behavior, predictable resource usage and reliable change propagation, preserving CDC guarantees.

> The results confirm that Debezium Server is not an experimental or limited runtime, but a well-engineered product built on top of Debezium's proven CDC engines.

Reviewer 2: ```suggestion
The results confirm that Debezium Server is not an experimental or limited runtime, but a well-engineered product built on top of Debezium's proven CDC engines powered by https://quarkus.io[Quarkus].
```

This will automatically create the EC2 instances, provision them with ansible (copy files and install docker) and start the scenario.
Logging in Grafana will allow you to see how everything evolves and progresses (inside Grafana you will need to navigate to the MySQL Streaming dashboard).

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/benchmark_completed_scenario.png" class="responsive-image" alt="Streaming dashboard from a completed scenario">
</div>
++++

With the technology stack clearly defined and reproducible, we can now focus on how performance will be measured and which metrics will be considered:

- CPU and RAM Percentage usage: We are monitoring the resource usage from the Debezium Server container.
- Network Input and Output: We are also taking into account the communications performed by the container.
- Disk Reads and Writes: To check how much disk space/bandwidth is needed.
- Throughput: This is the main value we are going to focus on (based on this value, we can estimate the operation latency) and we are comparing this to MySQL throughput. 

> For all the given metrics, we are calculating the average over the duration of the test. This way, we provide a stable value, and any possible spikes that may exist do not affect

Reviewer 3: ```suggestion
For each of the preceding metrics, we calculate the average value over the duration of the test. Calculating the average value in this way, we arrive at a stable value, so that any spikes in the data do not affect
```

the final result and are included in it. These are a few more detailed panels of the dashboard.

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/cpu_memory_consumption.png" class="responsive-image" alt="CPU and Memory consumption during one execution">
</div>
++++

In the picture above, you can see a few of the panels related to the resource consumption. The two on top are related to the CPU usage, which is the calculated percentage of usage for a 4 vCPUs instance.
And the other ones are the memory consumption in MB throughout the execution. The two on the left represent the evolution over time, and the ones on the right are the all-time average. 

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/throughput_time_evolution.png" class="responsive-image" alt="Throughput evolution during one execution">
</div>
++++

> In this example, we see the Debezium Server throughput from another execution of the test. This throughput chart is split into two parts, which comes from YCSB behavior.

Reviewer 3: ```suggestion
The preceding example shows the Debezium Server throughput that the test captures during execution. This throughput chart is split into two parts, a result that reflects the two phases of the YCSB test.
```

YCSB executions are divided into two phases: an initial load phase that populates the tables, followed by a workload phase that executes the configured mix of operations. This behavior explains the two distinct phases visible in the throughput charts.

> If the first half keeps the throughput at the set level, we consider the execution successful (OK/1), otherwise if something in the process makes it fail we consider it a failed execution (NOK/0).

Reviewer 3: ```suggestion
If during first phase we observe that the throughput remains consistent with the configured level, we consider the execution successful (OK/1). On the other hand, if something in the process results in variable throughput, we consider that the execution failed (NOK/0).
```

The image that follows shows a detailed view of the second phase of the execution, which runs 300 updates per second and 300 inserts per second. 

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/throughput_detailed_view.png" class="responsive-image" alt="Detailed view of the throughput panel">
</div>
++++

[id=results-analysis]
== Results and Analysis

Let's now examine the results to gain a clear view of how the system behaves under varying levels of load, and

> to identify trends that might not be evident from isolated measurements. If you want to review the detailed analysis, you can view the raw results [ADD_REPO_URL_HERE_RESULTS_FILE]. 

Reviewer 3: ```suggestion
Let's now examine the results to gain a clear view of how the system behaves under varying levels of load, so that we can identify trends that might not be evident from isolated measurements. If you want to review the detailed analysis, you can view the raw results [ADD_REPO_URL_HERE_RESULTS_FILE]. 
```

</div>
++++

With the preceding image in mind, let's consider the meaning of each column.

- TABLE_RECORDS: Desired table records we are going to work with in the execution. 
- TABLE_TARGET_OPS: Desired operations per second we are going to perform during the execution.
- TOTAL_RECORDS: Calculated value, the value of TABLE_RECORDS times the value of TABLES.
- TOTAL_OPS: Calculated value, the value of TABLE_TARGET_OPS times the value of TABLES.
- RESULT: Indicates if the execution has ended positively or not. 1 means OK and 0 means NOK.
- DURATION: Total duration of the execution. This includes both phases of YCSB, as explained above.
- AVG_*: The average (throughout the execution) of the given resource. Resources: CPU, MEM, NET_IN, NET_OUT, DISK_WRITE and DISK_READ.
Averaging smooths short-lived spikes and highlights sustained behavior, making it suitable for capacity planning but less precise for analyzing tail latency or transient saturation.

There are still a few hidden columns.

- DATABASE: The database picked to run the test against. In this case it is always MySQL.

> - TABLES: The amount of different tables we are using in this execution.

Reviewer 3: ```suggestion
- TABLES: The number of different tables we are using in this execution.
```
Should this say "The _number_ of different tables ...", or does this column list actual table names?

> - DISTRIBUTION: The distribution selected for the given execution. In all these examples it is always uniform. Go to YCSB options to select your preferred one.

Reviewer 3: ```suggestion
- DISTRIBUTION: The distribution selected for the given execution. The examples cited in this post all assume that the data distribution is set to` uniform`. You can modify the setting in the YCSB options.
```

> - COMMENTS: Place to write down any observation done during the execution.

Reviewer 3: ```suggestion
- COMMENTS: A field for recording observations taken while executing the test.
```

Side note: these are only the values for Debezium Server container. We are not keeping the data for other containers, although we can see them in the dashboard.

> Below is the analysis of the results obtained from the performance executions, whose objective is to objectively evaluate the system's behavior under different load

Reviewer 3: ```suggestion
The following analysis shows the results obtained from the performance executions, where the objective is to evaluate the system's behavior under different load
```

> scenarios and operating conditions. This analysis focuses on identifying trends, bottlenecks and relevant variations in the key metrics presented before.

Reviewer 3: ```suggestion
scenarios and operating conditions. This analysis focuses on identifying trends, bottlenecks, and relevant variations in the key metrics presented earlier.
```

Let us look at the charts and analyze the results.

> This chart presents the execution time (measured in hours) based on the operations per second and the number of tables involved in the execution.

Reviewer 3: ```suggestion
The chart that follows shows the execution time (measured in hours) based on the number of operations per second and the number of tables involved in the execution.
```

> This chart does not follow a single uniform growth pattern solely driven by the number of operations. While some table groups exhibit an increase in duration as total operations grow,

Reviewer 3: ```suggestion
The chart reveals that execution time does not follow a single uniform growth pattern that is driven solely by the number of operations. Although some table groups exhibit an increase in duration as total operations grow, but 
```
Per [IBM style,](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#word-usage__w) _"Use **while** only to refer to a period of time. To avoid potential ambiguity, do not use the term as a synonym for **although** or **though** in documentation."_

this behavior is not consistent across all tables. Several tables show relatively high execution times even at lower operation counts, whereas others maintain low durations despite handling more operations.
This indicates that execution time is strongly influenced by table-specific factors rather than by workload size alone.

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/duration_by_ops_table.png" class="responsive-image" alt="Duration by operation per second and table amount">
</div>
++++

> This chart shows the average CPU utilization. It does not scale linearly with the number of operations and varies across tables.

Reviewer 3: ```suggestion
The next chart shows average CPU utilization. Notice that rather than scaling linearly with the number of operations, utilization varies across tables.
```

In the initial table groups, CPU increases as the workload grows, reaching peak values at a mid-range operation count, after which it stabilizes or even decreases despite higher numbers of operations. 
For subsequent tables, CPU usage remains within a relatively narrow range, with only moderate fluctuations, indicating that higher workloads do not necessarily translate into proportionally higher CPU consumption.

CPU utilization is influenced more by table-specific processing characteristics and execution efficiency than by total operation volume alone.

> Other factors - I/O wait, batching or internal optimizations - may play a more significant role in performance at higher workloads. 

Reviewer 3: ```suggestion
Other factors - I/O wait, batching or internal optimizations - might play a more significant role in performance at higher workloads. 
```
Per IBM style,  _"Use [**may**] only to indicate permission. Otherwise, use **can** to indicate ability, or use **might** to indicate possibility. **May** is overused in technical information. In most cases, use **can** or **might** instead."_

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/cpu_by_ops_table.png" class="responsive-image" alt="CPU usage by operation per second and table amount">
</div>
++++

> The memory consumption remains stable across different operation counts and tables, with no clear relationship to the number of operations executed. Most measurements cluster within a similar range, suggesting a consistent baseline memory footprint largely independent of workload size.

Reviewer 3: ```suggestion
Memory consumption remains stable across different operation counts and tables, with no clear relationship to the number of operations executed. Most measurements cluster within a similar range, suggesting a consistent baseline memory footprint that is largely independent of workload size.
```

> The chart suggests that memory usage is dominated by logical database schema or configuration-related factors rather than by the volume of operations, and that memory is unlikely to be the primary scaling constraint. 

Reviewer 3: ```suggestion
The following chart suggests that memory usage is dominated by factors related to logical database schema or to configuration, rather than being directly related to the volume of operations. The analysis indicates that memory is unlikely to be the primary scaling constraint. 
```

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/memory_by_ops_table.png" class="responsive-image" alt="Memory usage by operation per second and table amount">
</div>
++++

> The network usage varies significantly across tables and operations counts, with output traffic consistently higher than input traffic in all cases. In the initial table groups, both input and output volumes increase as the number of operations grows, indicating a clear workload-driven effect on network utilization.

Reviewer 3: ```suggestion
As we can see in the next figure, network usage varies significantly across tables and operations counts, with output traffic being consistently higher than input traffic in all cases. In the initial table groups, both input and output volumes increase as the number of operations grows, indicating a clear relationship between workload and network utilization.
```

> However, this pattern becomes less consistent in later tables, where network usage remains high or fluctuates despite changes in total operations. This suggests that data transfer is influenced more by table-specific characteristics than by the number of operations.

Reviewer 3: ```suggestion
However, this pattern becomes less consistent in later tables, where network usage remains high or fluctuates despite changes in total operations. This finding suggests that data transfer is influenced more by table-specific characteristics than by the number of operations.
```

> Overall, this indicates that network utilization is a dominant and variable factor in system behavior, particularly on the output side, and should be considered a potential constraint depending on table structure and data change patterns. 

Reviewer 3: ```suggestion
Overall, we can infer that network utilization is a dominant and variable factor in system behavior, particularly on the output side, and can represent a potential constraint depending on table structure and data change patterns. 
```

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/network_by_ops_table.png" class="responsive-image" alt="Network usage by operation per second and table amount">
</div>
++++

> Disk write activity remains relatively stable across most tables and operation counts, indicating a consistent write pattern independent of workload size.

Reviewer 3: ```suggestion
Our next chart shows how disk activity varies with different table and operation counts. The chart indicates that disk write activity remains relatively stable across most tables and operation counts, indicating a consistent write pattern independent of workload size.
```

> In contrast, disk read activity is minimal or negligible in the early table groups and becomes more prominent only in later tables, where noticeable spikes appear at specific operation counts.

Reviewer 3: ```suggestion
By contrast, disk read activity is minimal or negligible in the early table groups, and becomes more prominent only in later tables, where noticeable spikes appear at specific operation counts.
```

One table in particular exhibits a pronounced write peak compared to all others, suggesting an isolated behavior such as increased flushing or checkpointing.
This behavior is typically observed during log file rotation or log exchanges, where disk activity is driven primarily by the number of log files generated and the overall execution duration, rather than by the volume of operations.

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/disk_by_ops_table.png" class="responsive-image" alt="Disk usage by operation per second and table amount">
</div>
++++

> The maximum throughput achieved varies noticeably across tables, indicating different upper performance limits depending on table characteristics.

Reviewer 3: ```suggestion
As we can see in the next figure, the maximum throughput that Debezium achieves varies noticeably across tables, indicating the effect that table characteristics can have on upper performance limits.
```

Tables 1 and 2 reach the highest values, demonstrating a greater capacity to sustain higher operation rates, while subsequent tables show a gradual reduction in maximum achievable throughput.

> This decrease is not strictly linear, as some tables outperform others despite their position, but it clearly reflects table-specific constraints such as data structure, indexing or processing complexity.

Reviewer 3: ```suggestion
This decrease is not strictly linear, as some tables outperform others despite their position, but it clearly reflects table-specific constraints, possibly related to data structure, indexing, or processing complexity.
```

> The chart highlights that maximum operations per second are primarily driven by table design and access patterns rather than by uniform system limits, reinforcing the need for table-level performance evaluation.

Reviewer 3: ```suggestion
The next chart highlights that the maximum number of operations per second is primarily driven by table design and access patterns, rather than by uniform system limits, reinforcing the importance of table-level performance evaluation.
```

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/max_ops_by_table.png" class="responsive-image" alt="Maximum operation per second by table amount">
</div>
++++

But, *what is Debezium Server's observed upper bound in number of operations per second?* 

> In this comparison we can see the relationship between MySQL and Debezium Server throughput. The initial peak observed at the start of the execution is the desired value we are targeting for this execution.

Reviewer 3: ```suggestion
In this next chart we compare the relationship between MySQL and Debezium Server throughput. The initial peak observed at the start of the execution matches the value that we are targeting for this execution.
```

1. *Start with Debezium Server and default configurations.* For small-to-medium projects, Debezium Server provides an excellent starting point without the operational complexity of Kafka Connect or Kubernetes.
The results show that default configurations already deliver stable, predictable and near real-time CDC performance, allowing teams to validate their use cases before investing time in advanced tuning or more complex architectures.
2. *Design tables and schemas with CDC in mind.* Performance is strongly influenced by table-level characteristics rather than raw operations counts. Careful schema design can have a significant impact on CDC throughput and latency.
Optimizing table structures will often yield greater benefits than tuning Debezium itself.
3. *Treat the database as the primary throughput limiter.* The benchmark shows that Debezium Server can keep up with MySQL in near real time and that overall throughput is bounded by database's write capacity.
Performance efforts should therefore focus first on the database before attempting to optimize the CDC layer.

> 4. *Scale architecture complexity only when necessary.* Kafka, Kubernetes and distributed CDC topologies are powerful but introduce non-trivial operational overhead. For lightweight or well-bounded workloads, Debezium Server alone is often sufficient and easier to operate.

Reviewer 3: ```suggestion
4. *Scale architectural complexity only when necessary.* Kafka, Kubernetes and distributed CDC topologies are powerful but introduce non-trivial operational overhead. For lightweight or well-bounded workloads, Debezium Server alone is often sufficient and easier to operate.
```

> More complex architectures should be adopted only when clear scalability, availability or integration requirements justify them. 

Reviewer 3: ```suggestion
Adopt more complex architectures only if the scalability, availability, or integration requirements clearly justify them. 
```

In summary, Debezium Server emerges as a simple, efficient and high-performance CDC solution that challenges the assumption that Kafka-centric or Kubernetes-based architectures are always required.

> For lightweight projects and streamlined deployments, it offers a compelling balance of maturity, simplicity and performance, delivering real-time change data capture while keeping infrastructure complexity to a minimum. 

Reviewer 3: ```suggestion
For lightweight projects and streamlined deployments, Debezium Server offers a compelling balance of maturity, simplicity and performance, delivering real-time change data capture while keeping infrastructure complexity to a minimum. 
```

[id=follow-up]
== What’s next?

> While the results presented above provide a clear view of Debezium Server behavior under default configurations, they also raise additional questions worth exploring in follow-up experiments:

Reviewer 3: ```suggestion
While the results presented in this post provide a clear view of Debezium Server behavior under default configurations, they also raise additional questions worth exploring in follow-up experiments, for example:
```

> - Discover Debezium Server limits: improve this experiments architecture and configurations to discover what is the point where Debezium Server cannot recover from it.

Reviewer 3: ```suggestion
- Discover Debezium Server limits: Improve the architecture and configuration of this experiment to discover the load limits for Debezium Server.
```
This item is a little bit unclear. How would improving the architecture and configurations in the preceding post help to reveal these limits? Is this a matter of _improving_ those, or more about modifying them?

> - Deploy Debezium with Kafka Connect: run this experiment with a distribuited and more complex architecture, using Kafka Connect to deploy Debezium.

Reviewer 3: ```suggestion
- Deploy Debezium with Kafka Connect: Run this experiment with a distributed and more complex architecture, using Kafka Connect to deploy Debezium.
```

> - Get involved with snapshotting: we have skiped over one important Debezium feature, and snapshotting deserves performance investigation.

Reviewer 3: ```suggestion
- Get involved with snapshotting: By omitting snapshots from our experiment, we skipped one the most important Debezium features. Snapshotting certainly merits further performance investigation.
```

- Run different sets of workloads and databases: we have run this experiment against MySQL and used a uniform workload. 

> Including of different databases, connectors and workload will give us the complete vision of Debezium capabilities. 

Reviewer 3: ```suggestion
- Run different sets of workloads and databases: We ran the present experiment against MySQL and used a uniform workload. 
For a more complete assessment of Debezium's capabilities, we should include different databases, connectors, and workloads in the test environment.
```

author: alvarvg
---

Performance is a critical concern when implementing Change Data Capture (CDC) solutions in production environments.
In this post, I am going to share how we have measured Debezium Server performance (deployed in a Docker container)
while streaming changes from MySQL to Kafka. All of this with the default configurations for Debezium, MySQL and Kafka.

+++<!-- more -->+++

This post does not aim to compare Debezium Server with Kafka Connect nor to identify absolute throughput limits. 
Instead, it focuses on understanding how Debezium Server behaves under realistic workloads using default configurations, 
and whether it can sustain real-time change data capture in lightweight architectures.

[id=introduction]
== Introduction

Measuring software performance is important. It directly affects reliability, scalability, cost efficiency and user trust. It is a prerequisite for making informed engineering decisions. 
Performance measurement transforms software from a black box into an understandable, predictable and optimizable system. It is well known that:

'What is not measured cannot be controlled, predicted or reliably improved'

> Going beyond sayings and proverbs, we can make the following assertions about performance.

Reviewer 3: ```suggestion
Going beyond sayings and proverbs, performance measurement provides you with evidence to inform and drive architectural decisions. 
Specifically, performance testing can help you by providing information that enables you to:
```

- Validates that the system meets the requirements
- Identifies bottlenecks and limiting factors
- Enables predictable scalability
- Supports data-driven architectural decisions
- Reduces operational and infrastructure costs
- Improves reliability and resilience

> - Builds trust with users and stakeholders

Reviewer 3: ```suggestion
- Validate that the system meets requirements
- Identify bottlenecks and limiting factors
- Help ensure predictable scalability
- Reduce operational and infrastructure costs
- Deliver improvements in reliability and resilience
- Build trust with users and stakeholders
```

[id=debezium-server]
== Debezium Server

Apart from performance tests, another great unknown in the CDC community is *Debezium Server*. If you are reading this, you probably know something about Debezium and CDC.
But let me introduce you to Debezium Server. 
Debezium Server is a standalone runtime for Change Data Capture that captures database changes using Debezium and delivers them directly to configurable sinks without requiring Kafka or Kafka Connect.
It addresses a common gap in CDC architectures: situations where Kafka Connect, Kubernetes, or large distributed platforms introduce unnecessary operational complexity. 

> It enables teams to adopt CDC with minimal infrastructure while still benefiting from Debezium’s mature and battle-tested connectors.

Reviewer 3: ```suggestion
It enables teams to adopt CDC with minimal infrastructure while still benefiting from Debezium’s mature and proven connectors.
```
Although I couldn't find official IBM or Red Hat guidelines, IMO, it's best to avoid terminology that refers to conflict as a metaphor. As one colleague put it: 

>among other reasons, there is enough real conflict in the world that some people are living with and suffering from; and also, considering the needs of a global audience and translation, it aids clarity to use a literal term rather than a metaphor.

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/debezium-server-architecture.png" style="max-width:70%;" class="responsive-image" alt="Debezium Server Architecture">
++++
*Debezium Server architecture*
====

Debezium Server is built on top of the Debezium Engine, packaging it into a production-ready application that can run as a single JVM process.
It runs as a single process and although Kafka is not required, it preserves the core CDC guarantees: 
- Exactly-once offset management (per connector)
- Ordered event delivery per table
- Snapshotting and streaming of the changes in the database
- Rich metadata about transactions and schemas

[id=benchmarking]
== Benchmarking

Now that we are familiar with all the components, let us dive into the technical details. 

Starting with the hardware, we have deployed two EC2 machines in AWS of type "c5a.xlarge". Each instance provides 4 vCPUs and 8 GiB of memory, ensuring sufficient resources for both workload generation and CDC processing.

> We have called them Observed and Observer (in order to separate the monitoring load from the machine where Debezium and MySQL are running). In the Observer instance, we have deployed Prometheus, Grafana and YCSB

Reviewer 3: ```suggestion
We have called these two instances _Observed_ and _Observer_ (in order to separate the monitoring load from the machine where Debezium and MySQL are running). In the Observer instance, we have deployed Prometheus, Grafana and YCSB
```

(the software piece generating the load). In the Observed instance, we have a single Kafka container (although it is not really necessary and another software can be used),

> Debezium Server and MySQL. 

Reviewer 3: ```suggestion
Debezium Server, and MySQL. 
```

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/benchmark_architecture_blueprint.png" style="max-width:70%;" class="responsive-image" alt="Benchmarking architecture">
++++
*Benchmarking architecture*
====

> As described above, we are using some software pieces, most of them are already known, except for YCSB. Yahoo! Cloud Serving Benchmark (YCSB), is a tool created by the Yahoo

Reviewer 3: ```suggestion
As per the preceding description, we are using several software pieces, most of which are already known, except for Yahoo! Cloud Serving Benchmark (YCSB). The YCSB tool, which was created by the Yahoo
```

team and released into open source as of 04/23/2010, developed to evaluate the performance of different cloud serving stores. This is the link to their GitHub project: https://ycsb.site[https://ycsb.site].

> If you want to reproduce this scenario, it is quite simple. Go to debezium-performance[ADD_REPO_URL_HERE] repo, clone it and cd into the infrastructure-automation folder and then run terraform apply. 

Reviewer 4: `[ADD_REPO_URL_HERE]` ?

Author: I told you about this one. Those are the placeholders to add the debezium performance repository. Is not yet ready to be published.

> If you want to reproduce this scenario, it is quite simple. Go to debezium-performance[ADD_REPO_URL_HERE] repo, clone it and cd into the infrastructure-automation folder and then run terraform apply. 

Reviewer 3: ```suggestion
If you want to reproduce this scenario, it is quite simple. Go to debezium-performance[ADD_REPO_URL_HERE] repo, clone it and cd into the infrastructure-automation folder and then run `terraform apply`. 
```
I am guessing that is a literal command for the user to run?

the final result and are included in it. These are a few more detailed panels of the dashboard.

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/cpu_memory_consumption.png" style="max-width:70%;" class="responsive-image" alt="CPU and Memory consumption during one execution">
++++
*CPU and Memory consumption during one execution*
====

In the picture above, you can see a few of the panels related to the resource consumption. The two on top are related to the CPU usage, which is the calculated percentage of usage for a 4 vCPUs instance.
And the other ones are the memory consumption in MB throughout the execution. The two on the left represent the evolution over time, and the ones on the right are the all-time average. 

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/throughput_time_evolution.png" style="max-width:70%;" class="responsive-image" alt="Throughput evolution during one execution">
++++
*Throughput evolution during one execution*
====

> In another example, we have the Debezium Server throughput during one of the executions. This throughput chart is split into two parts, which comes from YCSB behavior.

Reviewer 3: ```suggestion
In this example, we see the Debezium Server throughput from another execution of the test. This throughput chart is split into two parts, which comes from YCSB behavior.
```

YCSB executions are divided into two phases: an initial load phase that populates the tables, followed by a workload phase that executes the configured mix of operations. This behavior explains the two distinct phases visible in the throughput charts.
If the first half keeps the throughput at the set level, we consider the execution successful (OK/1), otherwise if something in the process makes it fail we consider it a failed execution (NOK/0).

> In the picture below, you can see a detailed view in the second phase of the execution where we are running 300 updates per second and 300 inserts per second. 

Reviewer 3: ```suggestion
The image that follows shows a detailed view of the second phase of the execution, which runs 300 updates per second and 300 inserts per second. 
```

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/throughput_detailed_view.png" style="max-width:70%;" class="responsive-image" alt="Detailed view of the throughput panel">
++++
*Detailed view of the throughput panel*
====

[id=results-analysis]
== Results and Analysis

> Let me now show you the results. By examining these results, it is possible to gain a clear view of how the system behaves under varying levels of load and

Reviewer 3: ```suggestion
Let's now examine the results to gain a clear view of how the system behaves under varying levels of load, and
```

> to identify trends that may not be evident from isolated measurements. But before getting into detailed analysis, you can find the raw results here[ADD_REPO_URL_HERE_RESULTS_FILE]. 

Reviewer 3: ```suggestion
to identify trends that might not be evident from isolated measurements. If you want to review the detailed analysis, you can view the raw results [ADD_REPO_URL_HERE_RESULTS_FILE]. 
```

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/debezium_server_streaming_mysql_results.png" style="max-width:70%;" class="responsive-image" alt="Debezium Server streaming results screenshot for MySQL">
++++
*Debezium Server streaming results screenshot for MySQL*
====

> With the image above in mind, let's understand what each of the columns means.

Reviewer 3: ```suggestion
With the preceding image in mind, let's consider the meaning of each column.
```

Conditions such as long-running transactions or other complex workload patterns are not represented,
as incorporating them would significantly increase the complexity of the test case.

[id=introduction]
== Introduction

Measuring software performance is important. It directly affects reliability, scalability, cost efficiency and user trust. It is a prerequisite for making informed engineering decisions. 
Performance measurement transforms software from a black box into an understandable, predictable and optimizable system. It is well known that:

'What is not measured cannot be controlled, predicted or reliably improved'

Going beyond sayings and proverbs, performance measurement provides you with evidence to inform and drive architectural decisions. 
Specifically, performance testing can help you by providing information that enables you to:

- Validate that the system meets requirements
- Identify bottlenecks and limiting factors
- Help ensure predictable scalability
- Reduce operational and infrastructure costs
- Deliver improvements in reliability and resilience
- Build trust with users and stakeholders

[id=debezium-server]
== Debezium Server

> Apart from performance tests, another great unknown in the CDC community is *Debezium Server*. If you are reading this, you probably know something about Debezium and CDC.

Reviewer: ```suggestion
Apart from performance tests, another great unknown in the CDC community might be *Debezium Server*. If you are reading this, you probably know something about Debezium and CDC.
```

But let me introduce you to Debezium Server. 
Debezium Server is a standalone runtime for Change Data Capture that captures database changes using Debezium and delivers them directly to configurable sinks without requiring Kafka or Kafka Connect.
It addresses a common gap in CDC architectures: situations where Kafka Connect, Kubernetes, or large distributed platforms introduce unnecessary operational complexity. 
It enables teams to adopt CDC with minimal infrastructure while still benefiting from Debezium’s mature and proven connectors.

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/debezium-server-architecture.png" style="max-width:70%;" class="responsive-image" alt="Debezium Server Architecture">
++++
*Debezium Server architecture*
====

Debezium Server is built on top of the Debezium Engine, packaging it into a production-ready application that can run as a single JVM process.
It runs as a single process and although Kafka is not required, it preserves the core CDC guarantees: 

> - Exactly-once offset management (per connector)

Reviewer: Empty line missing, so list is not rendered as list

> - Exactly-once offset management (per connector)

Reviewer: ```suggestion
- At-least-once offset management (per connector)
```

> - Ordered event delivery per table

Reviewer: This depends on sink. For some of them it is total ordering, for others only based on key.

We have called these two instances _Observed_ and _Observer_ (in order to separate the monitoring load from the machine where Debezium and MySQL are running). In the Observer instance, we have deployed Prometheus, Grafana and YCSB
(the software piece generating the load). In the Observed instance, we have a single Kafka container (although it is not really necessary and another software can be used),
Debezium Server, and MySQL. 

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/benchmark_architecture_blueprint.png" style="max-width:70%;" class="responsive-image" alt="Benchmarking architecture">
++++
*Benchmarking architecture*
====

As per the preceding description, we are using several software pieces, most of which are already known, except for Yahoo! Cloud Serving Benchmark (YCSB). The YCSB tool, which was created by the Yahoo
team and released into open source as of 04/23/2010, developed to evaluate the performance of different cloud serving stores. This is the link to their GitHub project: https://ycsb.site[https://ycsb.site].

> If you want to reproduce this scenario, it is quite simple. Go to debezium-performance[ADD_REPO_URL_HERE] repo, clone it and cd into the infrastructure-automation folder and then run terraform apply. 

Reviewer: ```suggestion
If you want to reproduce this scenario, it is quite simple. Go to debezium-performance[ADD_REPO_URL_HERE] repo, clone it and cd into the infrastructure-automation folder and then run `terraform apply`. 
```

If the first half keeps the throughput at the set level, we consider the execution successful (OK/1), otherwise if something in the process makes it fail we consider it a failed execution (NOK/0).

The image that follows shows a detailed view of the second phase of the execution, which runs 300 updates per second and 300 inserts per second. 

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/throughput_detailed_view.png" style="max-width:70%;" class="responsive-image" alt="Detailed view of the throughput panel">
++++
*Detailed view of the throughput panel*
====

[id=results-analysis]
== Results and Analysis

Let's now examine the results to gain a clear view of how the system behaves under varying levels of load, and
to identify trends that might not be evident from isolated measurements. If you want to review the detailed analysis, you can view the raw results [ADD_REPO_URL_HERE_RESULTS_FILE]. 

[.centered-image.responsive-image]
====
++++

> <img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/debezium_server_streaming_mysql_results.png" style="max-width:70%;" class="responsive-image" alt="Debezium Server streaming results screenshot for MySQL">

Reviewer: Would it be possible to use engineer notation like 1 M instead of the scientific one?

For subsequent tables, CPU usage remains within a relatively narrow range, with only moderate fluctuations, indicating that higher workloads do not necessarily translate into proportionally higher CPU consumption.

CPU utilization is influenced more by table-specific processing characteristics and execution efficiency than by total operation volume alone.
Other factors - I/O wait, batching or internal optimizations - may play a more significant role in performance at higher workloads. 

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/cpu_by_ops_table.png" style="max-width:70%;" class="responsive-image" alt="CPU usage by operation per second and table amount">
++++
*CPU usage by operation per second and table amount*
====

The memory consumption remains stable across different operation counts and tables, with no clear relationship to the number of operations executed. Most measurements cluster within a similar range, suggesting a consistent baseline memory footprint largely independent of workload size.

> The chart suggests that memory usage is dominated by structural or configuration-related factors rather than by the volume of operations, and that memory is unlikely to be the primary scaling constraint. 

Reviewer: Logical database schema stored in memory

The chart highlights that maximum operations per second are primarily driven by table design and access patterns rather than by uniform system limits, reinforcing the need for table-level performance evaluation.

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/max_ops_by_table.png" style="max-width:70%;" class="responsive-image" alt="Maximum operation per second by table amount">
++++
*Maximum operation per second by table amount*
====

But, *what is Debezium Server's observed upper bound in number of operations per second?* 

In this comparison we can see the relationship between MySQL and Debezium Server throughput. The initial peak observed at the start of the execution is the desired value we are targeting for this execution.
MySQL fails to keep pace at such high values (1500 operations per second in 1 table). On the Debezium side, the throughput closely mirrors MySQL activity. 
This correlation confirms that the CDC pipeline is processing database changes efficiently and in near real time, with no apparent backlog or loss. 

> Overall throughput behavior is driven by the database write pattern rather than by downstream processing constraints.

Reviewer: Would be good to capture QueueRemainingCapacity JMX metric to see/confirm the pattern - slow reads or writes?
